tools/AzureRM.BootStrapper/AzureRM.Bootstrapper.psm1

```diff
@@ -13,44 +13,124 @@ else
+    }
+    return $ProfileCache
+  }
+
+  $ProfileCache = "$HOME/.config/Microsoft/AzurePowerShell/ProfileCache" 
   return $ProfileCache
 }
 
-# Make Rest-Call
-function Get-RestResponse
+# Function to find the latest profile map from cache
+function Get-LatestProfileMapPath
+{
+  $ProfileCache = Get-ProfileCachePath
+  if (-Not (Test-Path $ProfileCache))
+  {
+    return
+  }
+
+  $ProfileMapPaths = Get-ChildItem $ProfileCache
+
+  if ($ProfileMapPaths -eq $null)
+  {
+    return
+  }
+
+  $LatestMapPath = $ProfileMapPaths[0]
+  foreach ($ProfileMapPath in $ProfileMapPaths)
+  {
+    if ($LatestMapPath.CreationTime -lt $ProfileMapPath.CreationTime)
+    {
+      $LatestMapPath = $ProfileMapPath
+    }
+  }
+  return $LatestMapPath
+}
+
+# Make Web-Call
+function Get-WebResponse
 {
-  $response = Invoke-RestMethod -ea SilentlyContinue -Uri $PSProfileMapEndpoint -ErrorVariable RestError -OutFile "$ProfileCache\ProfileMap.Json"
-  return $RestError
+  $WebResponse =  Invoke-WebRequest -uri $PSProfileMapEndpoint 
+  return $WebResponse
 }
 
 # Get-ProfileMap from Azure Endpoint
 function Get-AzureProfileMap
 {
   $ProfileCache = Get-ProfileCachePath
+
+  # If profile cache directory does not exist, create one.
   if(-Not (Test-Path $ProfileCache))
   {
     New-Item -ItemType Directory -Force -Path $ProfileCache | Out-Null
   }
 
-  $RestError = Get-RestResponse  
+  # Get online profile data using Web request
+  $WebResponse = Get-WebResponse  
 
-  if ($RestError)
+  # Get ETag value for OnlineProfileMap
+  $OnlineProfileMapETag = $WebResponse.Headers["ETag"]
+
+  # If profilemap json exists, compare online Etag and cached Etag; if not different, don't replace cache.
+  $LatestProfileMapPath = Get-LatestProfileMapPath
+  if ($LatestProfileMapPath -ne $null)
   {
-    $HttpStatusCode = $RestError.ErrorRecord.Exception.Response.StatusCode.value__
-    $HttpStatusDescription = $RestError.ErrorRecord.Exception.Response.StatusDescription
-    
-	Throw "Http Status Code: $($HttpStatusCode) `nHttp Status Description: $($HttpStatusDescription)"
+    [string]$ProfileMapETag = [System.IO.Path]::GetFileNameWithoutExtension($LatestProfileMapPath)
+    if (($ProfileMapETag -eq $OnlineProfileMapETag) -and (Test-Path $LatestProfileMapPath.FullName))
+    {
+      $ProfileMap = Get-Content -Raw -Path $LatestProfileMapPath.FullName -ErrorAction SilentlyContinue | ConvertFrom-Json 
+      return $ProfileMap
+    }
   }
 
-  $ProfileMap = Get-Content -Raw -Path "$ProfileCache\ProfileMap.json" -ErrorAction SilentlyContinue | ConvertFrom-Json 
-
-  return $ProfileMap
+  # If profilemap json doesn't exist, or if online hash and cached hash are different, cache online profile map
+  $ChildPathName = ($OnlineProfileMapETag) + ".json"
+  $CacheFilePath = (Join-Path $ProfileCache -ChildPath $ChildPathName)
+  $OnlineProfileMap = RetrieveProfileMap -WebResponse $WebResponse
+  $OnlineProfileMap | ConvertTo-Json -Compress | Out-File -FilePath $CacheFilePath
```

Comment:
We need to keep Old ProfileMaps in Cache in case the user has a profile installed from that profilemap. So, when user asks "RemovePreviousVersions", we will use the old profile maps as a reference to uninstall previous versions.

Also, we will not be updating profile maps. If there is a different one available online, we will simply create a new one in cache. We might need file locks in case there are parallel requests to read the profile map.